shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt

```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
+        // that only one of those routes is typical, in which case we don't want to display it as a
+        // grouped direction in the UI. If there are only predicted trips on a single route, this
+        // will split the grouped direction up into as many headsign rows as there are headsigns.
+        fun resolveRealtimePatternForDirection(
+            staticData: NearbyStaticData.StaticPatterns.ByDirection,
+            upcomingTripsMap: UpcomingTripsMap?,
+            parentStopId: String,
+            alerts: Collection<Alert>?,
+            hasSchedulesTodayByPattern: Map<String, Boolean>?,
+            allDataLoaded: Boolean,
+        ): List<RealtimePatterns> {
+            val typicalPatternsByRoute =
+                staticData.patterns
+                    .filter { it.typicality == RoutePattern.Typicality.Typical }
+                    .groupBy({ it.routeId }, { it.id })
+
+            // If data hasn't loaded, or there are more than one typical routes in this direction,
+            // we never want to split into individual headsign rows.
+            if (!allDataLoaded || typicalPatternsByRoute.size > 1) {
+                return listOf(
+                    RealtimePatterns.ByDirection(
+                        staticData,
+                        upcomingTripsMap,
+                        parentStopId,
+                        alerts,
+                        hasSchedulesTodayByPattern,
+                        allDataLoaded
+                    )
+                )
+            }
+
+            val patternsById = staticData.patterns.associateBy { it.id }
+            val tripsByPattern =
+                upcomingTripsMap
+                    .orEmpty()
+                    .mapNotNull {
+                        when (val key = it.key) {
+                            is RealtimePatterns.UpcomingTripKey.ByRoutePattern ->
+                                if (
+                                    key.routePatternId != null &&
+                                        patternsById.containsKey(key.routePatternId) &&
+                                        key.parentStopId == parentStopId
+                                ) {
+                                    key.routePatternId to it.value
+                                } else {
+                                    null
+                                }
+                            else -> null
+                        }
+                    }
+                    .toMap()
+            val upcomingPatternsByRouteAndHeadsign =
+                tripsByPattern.values
+                    .flatten()
+                    .groupBy({ it.trip.routeId }, { it.trip.headsign to it.trip.routePatternId })
+                    .mapValues { routeEntry ->
+                        routeEntry.value.groupBy({ it.first }, { it.second })
+                    }
+
+            val predictedPatternsByRoute =
```

Comment:
question (non-blocking): Confirming my own learning - if we skip the step of checking predicted patterns by route, then we'd break down into showing individual headsigns prematurely because there could be some late-at-night atypical headsigns.